validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,162 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
+)
+
+func (c beaconApiValidatorClient) waitForActivation(ctx context.Context, in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	// Represents the set of keys in `in`
+	targetPubKeys := make(map[string]struct{})
+
+	// Represents the set of keys actually returned by the beacon node
+	retrievedPubKeys := make(map[string]struct{})
+
+	// Contains all keys in targetPubKeys but not in retrievedPubKeys
+	missingPubKeys := make(map[string]struct{})
+
+	statuses := []*ethpb.ValidatorActivationResponse_Status{}
+
+	for _, publicKey := range in.PublicKeys {
+		stringPubKey := "0x" + hex.EncodeToString(publicKey)
+		targetPubKeys[stringPubKey] = struct{}{}
+	}
+
+	params := url.Values{}
+
+	for _, publicKey := range in.PublicKeys {
+		params.Add("id", fmt.Sprintf("0x%s", hex.EncodeToString(publicKey)))
+	}
```

Comment:
You already did this string encoding once. Now you can use the `targetPubKeys` map keys to set up these params while reusing the encoding work that was already done previously. 

```suggestion
	for k := range targetPubKeys {
		params.Add("id", k)
	}
```

Reply:
Yes, or I can iterate only once over `in.PublicKeys`.